Nikon mount - Sigma Lens Updates

frli8848 opened the pull request:
This is my first GitHub/Exiv2 contribution (attempt) and it adds:
* Sigma 20mm F1.4 DG HSM | A : mount/UPC code,
* Sigma 50mm F1.4 DG HSM | A : mount/UPC code,
* Sigma 120-300mm F2.8 DG OS HSM | S info/support (lens Id etc.).

/Fredrik

Commit:
Added Sigma 20mm F1.4 DG HSM | A mount/UPC code (for Nikon mount).

frli8848 committed:
Added Sigma 50mm F1.4 DG HSM | A mount/UPC code (for Nikon mount).

frli8848 committed:
Added the Sigma 120-300mm F2.8 DG OS HSM | S for Nikon mount.

clanmills commented:
Thank You, Fredrik.  Can you attach sample images with which we can test your changes.  We normally extract the metadata from your images and into an .exv file and add that to our test suite to ensure your changes are disturbed by future changes.

In v0.26, I added a configuration file ~/.exiv2 to enable you to change lens definitions on your computer without modifying code.  http://dev.exiv2.org/issues/1336

frli8848 commented:
Hi Robin,

The Sigma product numbers I got from Sigma's lens brochure:

  http://www2nd.sigma-photo.co.jp/downloads/catalog/SIGMA-lens-En.pdf

Addition of these number should not break anything.

And here is a jpg from a Sigma 120-300 DG OS HSM | S (Sport) lens: 
![_dsc0246](https://user-images.githubusercontent.com/36279125/36028018-c69cfd3c-0d9d-11e8-97d2-cc1b35fc5aa6.JPG)

BTW: Did I do the pull request in the preferred way? The instructions at the Redmine wiki

http://dev.exiv2.org/projects/exiv2/wiki/Creating_a_PR_on_GitHub

mentions creating a branch for the pull request while the CONTRIBUTING.md file at:

https://github.com/Exiv2/exiv2/blob/master/CONTRIBUTING.md

do not!?

https://github.com/Exiv2/exiv2/blob/master/CONTRIBUTING.md

clanmills commented:
Fredrik

Thanks for sending this.  The instructions in CONTRIBUTING.md are correct and I'll fix the instructions on Redmine later.

I believe your patch is good and will work perfectly.  I've built/tested using your test file:
```
592 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ exiv2 -pa --grep lens/i ~/Downloads/36028018-c69cfd3c-0d9d-11e8-97d2-cc1b35fc5aa6.JPG 
Exif.Nikon3.LensType                         Byte        1  D G VR
Exif.Nikon3.Lens                             Rational    4  120-300mm F2.8
Exif.Nikon3.LensFStops                       Undefined   4  6
Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 120-300mm F2.8 DG OS HSM | S
Exif.NikonLd3.LensFStops                     Byte        1  F6.0
593 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ 
```

**_May I request test images for each lens, please_**?  It's very common for the same LensID to be used for multiple lenses.  For this reasons, it's very desirable to have a test in our test suite for your lens.  It's possible for a future PR to disturb your patch!   

@D4N is a Git expert and will probably comment on your PR.  I'm going to leave merging your code and updating of the test harness to him as we have a work-in-progress project to re-engineer our test harness.  If I submit your code and test image, I'm confident he will say I've done it the wrong way!

**_Have you tried your lens/.exiv2?_**  It works fine for the image you've provided.

```
583 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/test $ exiv2 -pa --grep lens/i ~/Downloads/36028018-c69cfd3c-0d9d-11e8-97d2-cc1b35fc5aa6.JPG 
Exif.Nikon3.LensType                         Byte        1  D G VR
Exif.Nikon3.Lens                             Rational    4  120-300mm F2.8
Exif.Nikon3.LensFStops                       Undefined   4  6
Exif.NikonLd3.LensIDNumber                   Byte        1  141
Exif.NikonLd3.LensFStops                     Byte        1  F6.0
584 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/test $ exiv2 -pv --grep lens/i ~/Downloads/36028018-c69cfd3c-0d9d-11e8-97d2-cc1b35fc5aa6.JPG 
0x0083 Nikon3       LensType                    Byte        1  14
0x0084 Nikon3       Lens                        Rational    4  1200/10 3000/10 28/10 28/10
0x008b Nikon3       LensFStops                  Undefined   4  72 1 12 0
0x000c NikonLd3     LensIDNumber                Byte        1  141
0x000d NikonLd3     LensFStops                  Byte        1  72
585 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/test $ bbedit "/Users/rmills/.exiv2"

586 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/test $ cat ~/.exiv2 
[nikon]
141=Fredrik's Sigma Sports Lens
587 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/test $ exiv2 -pa --grep lens/i ~/Downloads/36028018-c69cfd3c-0d9d-11e8-97d2-cc1b35fc5aa6.JPG 
Exif.Nikon3.LensType                         Byte        1  D G VR
Exif.Nikon3.Lens                             Rational    4  120-300mm F2.8
Exif.Nikon3.LensFStops                       Undefined   4  6
Exif.NikonLd3.LensIDNumber                   Byte        1  Fredrik's Sigma Sports Lens
Exif.NikonLd3.LensFStops                     Byte        1  F6.0
588 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/test $ 
```

frli8848 commented:
Robin, 

* Here is a jpg from a Sigma 20mm F1.4 DG HSM | A (Nikon mount)
![_dsc2057_sigma 20mm_f1 4_dg_hsm_a](https://user-images.githubusercontent.com/36279125/36063583-0947187a-0e7f-11e8-9872-1e00c2e9b060.JPG)

* Here is a jpg from a Sigma 50mm F1.4 DG HSM | A (Nikon mount)
![_dsc2059_sigma 50mm_f1 4_dg_hsm_a](https://user-images.githubusercontent.com/36279125/36063593-3482bc92-0e7f-11e8-8777-b378b20a0b29.JPG)

And here is the output from exiv2 master from yesterday with my commits:
<pre>
fl@nisse ~/bilder/exiv2 $ exiv2 -pa --grep lens/i *.JPG
_DSC2057_Sigma 20mm_F1.4_DG_HSM_A.JPG  Exif.Nikon3.LensType                         Byte        1  D G 
_DSC2057_Sigma 20mm_F1.4_DG_HSM_A.JPG  Exif.Nikon3.Lens                             Rational    4  20mm F1.4
_DSC2057_Sigma 20mm_F1.4_DG_HSM_A.JPG  Exif.Nikon3.LensFStops                       Undefined   4  7
_DSC2057_Sigma 20mm_F1.4_DG_HSM_A.JPG  Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 20mm F1.4 DG HSM | A
_DSC2057_Sigma 20mm_F1.4_DG_HSM_A.JPG  Exif.NikonLd3.LensFStops                     Byte        1  F7.0
_DSC2059_Sigma 50mm_F1.4_DG_HSM_A.JPG  Exif.Nikon3.LensType                         Byte        1  D G 
_DSC2059_Sigma 50mm_F1.4_DG_HSM_A.JPG  Exif.Nikon3.Lens                             Rational    4  50mm F1.4
_DSC2059_Sigma 50mm_F1.4_DG_HSM_A.JPG  Exif.Nikon3.LensFStops                       Undefined   4  7
_DSC2059_Sigma 50mm_F1.4_DG_HSM_A.JPG  Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 50mm F1.4 DG HSM | A
_DSC2059_Sigma 50mm_F1.4_DG_HSM_A.JPG  Exif.NikonLd3.LensFStops                     Byte        1  F7.0
</pre>

clanmills commented:
Thanks for sending this, Fredrik.  I have your test images safely stored on our buildserver and I've integrated your code into my fork http://github.com/clanmills/exiv2

Everything looks fine and I'll leave it to Dan to integrate your patch and update the new test harness.

```
501 rmills@rmillsmbp:~ $ exiv2 -pa --grep lens/i http://exiv2.dyndns.org:8080/userContent/testfiles/227/227-Sports.jpg
Exif.Nikon3.LensType                         Byte        1  D G VR
Exif.Nikon3.Lens                             Rational    4  120-300mm F2.8
Exif.Nikon3.LensFStops                       Undefined   4  6
Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 120-300mm F2.8 DG OS HSM | S
Exif.NikonLd3.LensFStops                     Byte        1  F6.0
502 rmills@rmillsmbp:~ $ exiv2 -pa --grep lens/i http://exiv2.dyndns.org:8080/userContent/testfiles/227/227-Sigma20.jpg
Exif.Nikon3.LensType                         Byte        1  D G 
Exif.Nikon3.Lens                             Rational    4  20mm F1.4
Exif.Nikon3.LensFStops                       Undefined   4  7
Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 20mm F1.4 DG HSM | A
Exif.NikonLd3.LensFStops                     Byte        1  F7.0
503 rmills@rmillsmbp:~ $ exiv2 -pa --grep lens/i http://exiv2.dyndns.org:8080/userContent/testfiles/227/227-Sigma50.jpg
Exif.Nikon3.LensType                         Byte        1  D G 
Exif.Nikon3.Lens                             Rational    4  50mm F1.4
Exif.Nikon3.LensFStops                       Undefined   4  7
Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 50mm F1.4 DG HSM | A
Exif.NikonLd3.LensFStops                     Byte        1  F7.0
504 rmills@rmillsmbp:~ $ 
```

clanmills commented:
Can you ask me this question another way.  I don't know what you're asking!

I didn't write Exiv2, although I've been working on the code for 10 years and know it quite well.  I've no idea how the Nikon Lens Code works.  You can see it credited to:  http://www.rottmerhusen.com (lens_id@rottmerhusen.com)

However, I'll figure out how it works and give you answer.  What's the question?
```
579 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ exiv2 -pa --grep lens/i test/tmp/reference/227-Sigma20.jpg 
Exif.Nikon3.LensType                         Byte        1  D G 
Exif.Nikon3.Lens                             Rational    4  20mm F1.4
Exif.Nikon3.LensFStops                       Undefined   4  7
Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 20mm F1.4 DG HSM | A
Exif.NikonLd3.LensFStops                     Byte        1  F7.0
580 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ exiv2 -pv --grep lens/i test/tmp/reference/227-Sigma20.jpg 
0x0083 Nikon3       LensType                    Byte        1  6
0x0084 Nikon3       Lens                        Rational    4  200/10 200/10 14/10 14/10
0x008b Nikon3       LensFStops                  Undefined   4  84 1 12 0
0x000c NikonLd3     LensIDNumber                Byte        1  121
0x000d NikonLd3     LensFStops                  Byte        1  84
581 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ 
```

clanmills commented:
The Nikon Lens code uses the LensIDNumber in the metadata, and that's what I use as the key in .exiv2 to enable you to redefine the Lens:
```
587 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ exiv2 -pv --grep LensIDNumber test/tmp/reference/227*.jpg
test/tmp/reference/227-Sigma20.jpg  0x000c NikonLd3     LensIDNumber                Byte        1  121
test/tmp/reference/227-Sigma50.jpg  0x000c NikonLd3     LensIDNumber                Byte        1  136
test/tmp/reference/227-Sports.jpg  0x000c NikonLd3     LensIDNumber                Byte        1  141
588 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ cat ~/.exiv2 
[nikon]
121=Frederik's 20mm Lens
141=Frederik's Sports Lens
136=Frederik's 50mm Lens
589 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ exiv2 -pa --grep LensIDNumber test/tmp/reference/227*.jpg
test/tmp/reference/227-Sigma20.jpg  Exif.NikonLd3.LensIDNumber                   Byte        1  Frederik's 20mm Lens
test/tmp/reference/227-Sigma50.jpg  Exif.NikonLd3.LensIDNumber                   Byte        1  Frederik's 50mm Lens
test/tmp/reference/227-Sports.jpg  Exif.NikonLd3.LensIDNumber                   Byte        1  Frederik's Sports Lens
590 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ 
```

In nikonmn_int.cpp, the table is indexed:  (0x79 = 121) and finds your string.
```
{0x79,0x54,0x31,0x31,0x0C,0x0C,0x4B,0x06,0x00,0x00,0x00, "Sigma", "", "20mm F1.4 DG HSM | A"},
591 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ exiv2 -pa --grep LensIDNumber test/tmp/reference/227-Sigma50.jpg
Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma 50mm F1.4 DG HSM | A
592 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ 
```
I don't know what all the data means in that table.  It's quite possibly not used by Exiv2.  However if you understand it and wish to change it, I'll be happy to accept your PR, provided it doesn't break our test suite.  However if we can't be certain that your change is beneficial and can be tested, I don't want to change code without certain purpose.  _If it's not broken, don't touch it!_ 

As you know, I really like to have a unique test in the test suite for every lens.  That could be a real challenge to find suitable images without purchasing 88 more lenses!  Incidentally, I have a Nikon 5300 with a Sigma 18-250mm (ID=146/0x92).  I also have Nikkor 12-24mm (ID=122/0x7a).  Clearly
```
{0x92,0x39,0x2D,0x88,0x2C,0x40,0x4B,0x0E,0x01,0x00,0x00, "Sigma", "", "18-250mm F3.5-6.3 DC OS Macro HSM"},
{0x7A,0x3C,0x1F,0x37,0x30,0x30,0x7E,0x06,0x01,0x01,0x00, "Nikon", "JAA784DA", "AF-S DX Zoom-Nikkor 12-24mm f/4G IF-ED"},
```

Clearly, the code knows the lens manufacturer and there by which table to index.  Perhaps LensType means that, however I'm guessing as I haven't read the code.

frli8848 commented:
I don't exactly now how exiv2 identifies a lens but my guess is that the lens ID and other parameters such as focal length, number of stops that the lens has, etc.,  are all used (in case of conflicting lens IDs, for example).  For the Nikon mounts these parameters are stored in the struct (starting at line 1824 in the nikonmn_int.cpp file):
<pre>
static const struct FMntLens {unsigned char lid,stps,focs,focl,aps,apl,lfw, ltype, tcinfo, dblid, mid; const char *manuf, *lnumber, *lensname;}
fmountlens[] = {
{0x01,0x58,0x50,0x50,0x14,0x14,0x02,0x00,0x00,0x00,0x00, "Nikon", "JAA00901", "AF Nikkor 50mm f/1.8"},
{0x01,0x58,0x50,0x50,0x14,0x14,0x05,0x00,0x00,0x00,0x00, "Nikon", "JAA00901", "AF Nikkor 50mm f/1.8"},
{0x02,0x42,0x44,0x5C,0x2A,0x34,0x02,0x00,0x00,0x00,0x00, "Nikon", "JAA72701", "AF Zoom-Nikkor 35-70mm f/3.3-4.5"},
{0x02,0x42,0x44,0x5C,0x2A,0x34,0x08,0x00,0x00,0x00,0x00, "Nikon", "JAA72701", "AF Zoom-Nikkor 35-70mm f/3.3-4.5"},

-snip-
</pre>

My question was about the 13:th parameter (item), <tt>*lnumber</tt>, in this struct. For Sigma lenses it is called Mount/UPC code by Sigma and can be found in their lens bruchure. And, I just wondered if is a good idea to add all "lnumbers" that I can find in Sigma's lens brochure to the <tt>FMntLens</tt> struct or not?

I have tried to figure out what the struct items are and my best guess :-) is 
* <tt>lid</tt> = Lens ID
* <tt>stps</tt> = Number of stops
* <tt>focs</tt> = Focal Short 
* <tt>focl</tt> = Focal Long (idenctical to focs on a prime lens)
* <tt>aps</tt> = Aperture @ Short 
* <tt>apl</tt> =  Aperture @ Long  (idenctical to aps on a prime lens)
* <tt>lfw</tt> = Lens Firmware 
* <tt>ltype</tt> = Lens Type 
* <tt>tcinfo</tt> = ?
* <tt>dblid</tt> = ?
* <tt>mid</tt> = ?
* <tt>manuf</tt> =  Manufacturer
* <tt>lnumber</tt> = Manufacturer lens number
* <tt>lensname</tt> = Lens name

clanmills commented:
The primary data used to find a lens is the LensID which is stored in the Manufacturer's MakerNote (which is different for every manufacturer).  Once the lens has been identified, if there is more than one lens with the same ID, code is used to inspect other metadata.  In effect, we guess the lens when there is possible ambiguity.  That's why, in my opinion, it's important to have a test image in the test harness.   Changes to the guessing code have to be careful not to break existing recognition.

The maintenance of the recognition code is laborious.  Last week I had 4 requests for new lenses.  So, for v0.26, I added the ~/.exiv2 mechanism to make it possible for users to "register" their lens.

Thanks for investigating the Nikon code.  As far as I can see, `lnumber` is ignored by the Exiv2 code.  I believe that data was imported directly from http://www.rottmerhusen.com and while other software using those data arrays may be interested in `lnumber`, I don't believe Exiv2 does anything with it.

```
598 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ grep lnumber src/*.cpp
src/nikonmn_int.cpp:static const struct FMntLens {unsigned char lid,stps,focs,focl,aps,apl,lfw, ltype, tcinfo, dblid, mid; const char *manuf, *lnumber, *lensname;}
599 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ 
``` 

So, that brings me to several reasons to change nothing:

1) It does no good
2) It's different from the data on http://www.rottmerhusen.com
3) We might unwittingly break something

It appears that the code on http://www.rottmerhusen.com has not been changed since 2011.  Perhaps you should work with Robert Rottmerhusen to bring that code up to date and I will import his "Latest and Greatest" when available.

If you'd like to be involved with the development of Exiv2, I would be very happy to accept your contributions.   https://github.com/Exiv2/exiv2/issues/225  If you'd like to discuss contributing with me one-to-one,  I'll be happy to talk on Skype or FaceTime or the phone.

frli8848 commented:
I sent  Robert Rottmerhusen an email and pointed him to this pull request. I will update here if/when I get a response.

/F

D4N commented:
I guess this is then on hold until Robert Rottmerhusen replies, right?

@frli8848 Please ping when you want to integrate this, I'll give you a hand with the test suite.

frli8848 commented:
I think it is safe to integrate my patch since it just adds support for the missing Sigma 120-300 sport lens and adds the lens numbers (<tt>lnumber</tt>) for the 20 mm and 50 mm 1.4 Art sigma lenses.  This is just a 3 line change of the code in the lens table struct.

I asked Robert Rottmerhusen for more details on what each element of the _extended_ lens ID means and how (and by whom) the Nikon (mount) exiv2 code is maintained.

And, I can get access to a few more Nikon and Sigma lenses. Is there a list of which lenses you have example images for in your test suite?

clanmills commented:
I agree that your patch is OK and should be integrated (with tests).

The close (however different) issue of using Robert Rottmerhusen's latest code can be raised as a different issue when we know that updated code is available.  I see he has been updating his web site in 2017.  I couldn't find the files that are included in Exiv2 on his web site.

I'm not too concerned about the meaning of the fields in his data tables.  I think the only one we use is `lid` (LensID).

We don't have a list of Lenses in the test suite, however you can run exiv2 -pa --grep lens/i test/data/* to find out what's there.  I've been looking after the 'lens recognition' code since Niels left the project about 3 years ago.  When I've modified lens code, I've asked the user to provide a test file from which I have extracted the '.exv' file and added it to the test suite.

I added the ~/.exiv2 feature in 0.26 to enable users to update lens recognition without having to wait for a code update from Exiv2.

frli8848 commented:
I have looked a little bit closer on which fields that is used by exiv2 to identify lenses (Nikon F-mounts only) and too me it looks like the first 8 ones are used. From line 2661 in <tt>src/nikonmn_int.cpp</tt>:
<pre>
-snip-

           if (   raw[0] == fmountlens[i].lid
                // stps varies with focal length for some Sigma zoom lenses.
                &&(raw[1] == fmountlens[i].stps || strcmp(fmountlens[i].manuf, "Sigma") == 0)
                && raw[2] == fmountlens[i].focs
                && raw[3] == fmountlens[i].focl
                && raw[4] == fmountlens[i].aps
                && raw[5] == fmountlens[i].apl
                && raw[6] == fmountlens[i].lfw
                && raw[7] == fmountlens[i].ltype) {
                // Lens found in database
                return os << fmountlens[i].manuf << " " << fmountlens[i].lensname;
            }

-snip-
</pre>
I think these must correspond to the 8 columns in the table at Robert Rottmerhusen web page:

http://www.rottmerhusen.com/objektives/lensid/thirdparty.html

I can also see that Darktable (which uses exiv2) displays the string composed of <tt>manuf</tt>+<tt>lensname</tt> so it looks like the code above is used (I have not tested this in a debugger to verify).

Some lenses have the same LensID but differ in other fields like, for example:
<pre>
-snip-

{0xA0,0x40,0x2D,0x53,0x2C,0x3C,0xCA,0x8E,0x01,0x00,0x03, "Nikon", "JAA826DA", "AF-P DX Nikkor 18-55mm f/3.5-5.6G VR"},
{0xA0,0x40,0x2D,0x53,0x2C,0x3C,0xCA,0x0E,0x01,0x00,0x03, "Nikon", "JAA827DA", "AF-P DX Nikkor 18-55mm f/3.5-5.6G"},

-snip-
</pre>

The <tt>tcinfo</tt>, <tt>dblid</tt>, and  <tt>mid</tt> fields doesn't seem to be used but they have non-zero values for some lenses.

clanmills commented:
This is really good stuff, Frederik.  You've studied and understood the Nikon MakerNote code.  Thank You for digging into this.

I'm always looking for volunteers to help with Exiv2.  Would you be willing to join us as "King MakerNotes".  Niels from Denmakr did this task for 3 or 4 years and was an excellent contributor.  He's gone back to collage to study (on top of his full time job) and left us.  I'm sure Niels will give you some mentoring and lots of encouragement (and so will I, Dan, Luis, Michal and Thomas).  Join Us?

a17r commented:
Seeing this, I have actually access to a lot of modern Sigma Art lenses for Canon and Nikon mount. (Plus others...). Anything I can do with that?

frli8848 commented:
I can probably help a little bit now and then but I'm rather busy at work and will be for most of the year. Regarding test files, can this just be as easy as adding another folder to the exiv2's git repo where, for example, (small) jpg files - one for each lens - is added? One can then add a (cmake) test which runs through all the test jpg:s and compare exiv2's output with the correct lens name string.

clanmills commented:
That's how the existing test harness works!  @D4N is re-engineering the test suite in python at the moment.  I believe he'll add tests for your three images above and that will give you the model to follow.  There's already something in src/Makefile to run the suite, so you can add a test image (jpg/exv) test/data and the test itself similar to tests/bugfixes/github/test_issue_208.py  It's better to add an exv file because it's smaller (an exv is a tiny jpeg with only metadata and no graphics.
```
987 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/tests/bugfixes/github $ cd /mmHD/Users/rmills/Jenkins/testfiles/227
rmills@rmillsmbp:/mmHD/Users/rmills/Jenkins/testfiles/227 $ ls -alt
total 8552
drwxr-xr-x  1 rmills  staff      264 10 Feb 16:16 .
-rw-r--r--@ 1 rmills  staff   582420 10 Feb 16:16 227-Sigma20.jpg
-rw-r--r--@ 1 rmills  staff   581888 10 Feb 16:16 227-Sigma50.jpg
-rw-r--r--@ 1 rmills  staff  3205873  9 Feb 16:07 227-Sports.jpg
drwxr-xr-x  1 rmills  staff      568  9 Feb 16:06 ..
rmills@rmillsmbp:/mmHD/Users/rmills/Jenkins/testfiles/227 $ exiv2 --verbose -ea *.jpg
File 1/3: 227-Sigma20.jpg
Writing Exif data from 227-Sigma20.jpg to ./227-Sigma20.exv
File 2/3: 227-Sigma50.jpg
Writing Exif data from 227-Sigma50.jpg to ./227-Sigma50.exv
File 3/3: 227-Sports.jpg
Writing Exif data from 227-Sports.jpg to ./227-Sports.exv
Writing XMP data from 227-Sports.jpg to ./227-Sports.exv
990 rmills@rmillsmbp:/mmHD/Users/rmills/Jenkins/testfiles/227 $ dir *.exv
-rw-r--r--  1 rmills  staff    20K 17 Feb 14:05 227-Sigma20.exv
-rw-r--r--  1 rmills  staff    20K 17 Feb 14:05 227-Sigma50.exv
-rw-r--r--  1 rmills  staff    25K 17 Feb 14:05 227-Sports.exv
rmills@rmillsmbp:/mmHD/Users/rmills/Jenkins/testfiles/227 $ 
```

D4N committed:
Added test for Sigma lenses

frli8848 commented:
@D4N I have not given the the repository github.com/Exiv2/exiv2.git any name - I just pressed the "Fork" button on GitHub, cloned my fork to disk, made some changes, comitted my changes, and pushed them to origin master of my fork. How do I tell my fork where  github.com/Exiv2/exiv2.git  is? I cannot find this info in the exiv2/CONTRIBUTING.md file.

D4N commented:
Run the following command in your forks directory, where you replace
$remote_name with a suitable name of your choice (origin is probably
already taken by your fork):

git remote add $remote_name https://github.com/Exiv2/exiv2.git

Then you should be able to pull from the official exiv2 repository.

Fredrik Lingvall <notifications@github.com> writes:

> @D4N I have not given the the repository github.com/Exiv2/exiv2.git any name - I just pressed the "Fork" button on GitHub, cloned my fork to disk, made some changes, comitted my changes, and pushed them to origin master of my fork. How do I tell my fork where  github.com/Exiv2/exiv2.git  is? I cannot find this info in the exiv2/CONTRIBUTING.md file.
>
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/227#issuecomment-367299691

frli8848 commented:
@D4N I have now done this:
<pre>
fl@nisse ~/projects/exiv2 $ git remote add exiv2_upstream https://github.com/Exiv2/exiv2.git
fl@nisse ~/projects/exiv2 $ git pull --rebase exiv2_upstream master
From https://github.com/Exiv2/exiv2
 * branch              master     -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Applying: Added Sigma 20mm F1.4 DG HSM | A mount/UPC code (for Nikon mount).
Applying: Added Sigma 50mm F1.4 DG HSM | A mount/UPC code (for Nikon mount).
Applying: Added the Sigma 120-300mm F2.8 DG OS HSM | S for Nikon mount.

fl@nisse ~/projects/exiv2 $ wget https://github.com/Exiv2/exiv2/files/1742251/0001-Added-test-for-Sigma-lenses.patch.gz

fl@nisse ~/projects/exiv2 $ gunzip 0001-Added-test-for-Sigma-lenses.patch.gz
fl@nisse ~/projects/exiv2 $ git am 0001-Added-test-for-Sigma-lenses.patch
fl@nisse ~/projects/exiv2 $ git push -f origin master
</pre>

D4N commented:
@frli8848 Thanks for your work!
@clanmills Any objections to merging this?

frli8848 commented:
@clanmills @D4N Can the steps I did here go into the exiv2/CONTRIBUTING.md file? Perhaps with the addition on how to generate the .exv files for the tests?

D4N commented:
Some of it can be surely added. What do you think is missing
specifically? The test suite stuff, the git commands or something else?

The generation of the .exv files should probably be added to
tests/doc.md or tests/writing_tests.md

Fredrik Lingvall <notifications@github.com> writes:

> @clanmills @D4N Can the steps I did here go into the exiv2/CONTRIBUTING.md file? Perhaps with the addition on how to generate the .exv files for the tests?
>
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/227#issuecomment-367308671

a17r commented:
> It would be wonderful to have lots of test files that we can run on both ExifTool and Exiv2 to ensure our compatibility.

I can definitely see to that once I got more spare time. I'm not much help with code, but providing test image files for all kinds of lenses where necessary could be my thing.

frli8848 commented:
@D4N For someone who is new to how GitHub works it would be nice to have a really simple (easy to follow) step-by-step instruction of the basic git commands to use - so that one just can copy-and-paste them to the terminal. That is, all steps from forking on GitHub to how we just push:ed the changes to get the pull request approved.

If there are other documents describing, for example, the test suite (generating test files etc) then there should be a pointer to those documents in the exiv2/CONTRIBUTING.md file. I think that the docs should be as self contained as possible to avoid linking to stuff that may change/brake things.

I also think it would be nice to have a "database" of lens (.exv) files that one (developers) can use to test code.  These files could be contributed also by non-programmers and the procedure to do this could also be described in the exiv2/CONTRIBUTING.md file. One can, for example, name the files like: <tt>nikon_f_mount_sigma_20mm_f1_4_dg_hsm_a.exv</tt>

clanmills commented:
@frli8848 @a17r About generating .exv files.  These are modified JPG file with no graphics, so they only contain metadata.  The normal way to create a .exv file is to extract it from a JPG `$ exiv2 -ea --verbose foo.jpg` will create the file foo.exv (which is typically about 20k).  ExifTool supports .exv files.

The test suite is stored in test/data and contains many .exv, .jpg files and a few .tif, .png and other formats.  Phil Harvey (of ExifTool) has about 1000+ JPG files and I'm sure will not object if we extract .exv files from his test suite to add to ours.  I recommend a courtesy email to Phil if we go down that road.

It's possible that there will be lenses not included in our test suite for which we can create a .exv using the exiv2 command-line program.  It would be great to have a test that compares our lens recognition to that achieved by ExifTool.  Such a test is of value to everybody - users, ExifTool and Exiv2.

If you are willing to undertake this project, I would be very happy to have an on-line meeting to discuss this.  We can talk on Skype or FaceTime or Google Hangouts.  I'm in England.  Conference Calls on Saturday or Sunday afternoon usually suit Exiv2 Contributors as it's the weekend and day-time almost everywhere except the Pacific Islands.  You can contact me directly at robin@clanmills.com

```
537 rmills@rmillsmbp:~/temp $ ls -alt ~/clanmills/Stonehenge.jpg
-rw-r--r--@ 1 rmills  staff  6757827 24 Nov  2016 /Users/rmills/clanmills/Stonehenge.jpg
538 rmills@rmillsmbp:~/temp $ ls -alth ~/clanmills/Stonehenge.jpg
-rw-r--r--@ 1 rmills  staff   6.4M 24 Nov  2016 /Users/rmills/clanmills/Stonehenge.jpg

539 rmills@rmillsmbp:~/temp $ exiv2 -ea --force --verbose http://clanmills.com/Stonehenge.jpg
File 1/1: http://clanmills.com/Stonehenge.jpg
Writing Exif data from http://clanmills.com/Stonehenge.jpg to Stonehenge.exv
Writing IPTC data from http://clanmills.com/Stonehenge.jpg to Stonehenge.exv
Writing XMP data from http://clanmills.com/Stonehenge.jpg to Stonehenge.exv

540 rmills@rmillsmbp:~/temp $ ls -alt Stonehenge.exv 
-rw-r--r--  1 rmills  staff  17989 21 Feb 13:55 Stonehenge.exv

541 rmills@rmillsmbp:~/temp $ exiv2 -pS Stonehenge.exv 
STRUCTURE OF JPEG FILE: Stonehenge.exv
 address | marker       |  length | data
       0 | 0xff01      
       7 | 0xffe1 APP1  |   15296 | Exif..II*......................
   15305 | 0xffe1 APP1  |    2610 | http://ns.adobe.com/xap/1.0/.<?x
   17917 | 0xffed APP13 |      68 | Photoshop 3.0.8BIM.......'.....
   17987 | 0xffd9 EOI  

542 rmills@rmillsmbp:~/temp $ exiv2 -pR Stonehenge.exv 
STRUCTURE OF JPEG FILE: Stonehenge.exv
 address | marker       |  length | data
       0 | 0xff01      
       7 | 0xffe1 APP1  |   15296 | Exif..II*......................
  STRUCTURE OF TIFF FILE (II): MemIo
   address |    tag                              |      type |    count |    offset | value
        10 | 0x010f Make                         |     ASCII |       18 |       146 | NIKON CORPORATION
        22 | 0x0110 Model                        |     ASCII |       12 |       164 | NIKON D5300
        34 | 0x0112 Orientation                  |     SHORT |        1 |           | 1
        46 | 0x011a XResolution                  |  RATIONAL |        1 |       176 | 300/1
        58 | 0x011b YResolution                  |  RATIONAL |        1 |       184 | 300/1
        70 | 0x0128 ResolutionUnit               |     SHORT |        1 |           | 2
        82 | 0x0131 Software                     |     ASCII |       10 |       192 | Ver.1.00 
        94 | 0x0132 DateTime                     |     ASCII |       20 |       202 | 2015:07:16 20:25:28
       106 | 0x0213 YCbCrPositioning             |     SHORT |        1 |           | 1
       118 | 0x8769 ExifTag                      |      LONG |        1 |           | 222
    STRUCTURE OF TIFF FILE (II): MemIo
     address |    tag                              |      type |    count |    offset | value
         224 | 0x829a ExposureTime                 |  RATIONAL |        1 |       732 | 10/4000
         236 | 0x829d FNumber                      |  RATIONAL |        1 |       740 | 100/10
         248 | 0x8822 ExposureProgram              |     SHORT |        1 |           | 0
         260 | 0x8827 ISOSpeedRatings              |     SHORT |        1 |           | 200
         272 | 0x8830 SensitivityType              |     SHORT |        1 |           | 2
         284 | 0x9000 ExifVersion                  | UNDEFINED |        4 |           | 0230
         296 | 0x9003 DateTimeOriginal             |     ASCII |       20 |       748 | 2015:07:16 15:38:54
         308 | 0x9004 DateTimeDigitized            |     ASCII |       20 |       768 | 2015:07:16 15:38:54
         320 | 0x9101 ComponentsConfiguration      | UNDEFINED |        4 |           | ...
         332 | 0x9102 CompressedBitsPerPixel       |  RATIONAL |        1 |       788 | 2/1
         344 | 0x9204 ExposureBiasValue            | SRATIONAL |        1 |       796 | 0/6
         356 | 0x9205 MaxApertureValue             |  RATIONAL |        1 |       804 | 43/10
         368 | 0x9207 MeteringMode                 |     SHORT |        1 |           | 5
         380 | 0x9208 LightSource                  |     SHORT |        1 |           | 0
         392 | 0x9209 Flash                        |     SHORT |        1 |           | 16
         404 | 0x920a FocalLength                  |  RATIONAL |        1 |       812 | 440/10
         416 | 0x927c MakerNote                    | UNDEFINED |     3152 |       820 | Nikon.....II*.....9.........0211 ...
      STRUCTURE OF TIFF FILE (II): MemIo
       address |    tag                              |      type |    count |    offset | value
            10 | 0x0001 Version                      | UNDEFINED |        4 |           | 0211
            22 | 0x0002 ISOSpeed                     |     SHORT |        2 |           | 0 200
            34 | 0x0004 Quality                      |     ASCII |        8 |       698 | NORMAL 
            46 | 0x0005 WhiteBalance                 |     ASCII |       13 |       706 | AUTO        
            58 | 0x0007 Focus                        |     ASCII |        7 |       720 | AF-A  
            70 | 0x0008 FlashSetting                 |     ASCII |       13 |       728 |             
            82 | 0x0009 GPSStatus                    |     ASCII |       20 |       742 |                    
            94 | 0x000b ProcessingSoftware           |    SSHORT |        2 |           | 0 0
           106 | 0x000c GPSSpeedRef                  |  RATIONAL |        4 |       762 | 538/256 354/256 256/256 256/256
           118 | 0x000d GPSSpeed                     | UNDEFINED |        4 |           | ...
           130 | 0x000e GPSTrackRef                  | UNDEFINED |        4 |           | ...
           142 | 0x0012 GPSMapDatum                  | UNDEFINED |        4 |           | ...
           154 | 0x0013 GPSDestLatitudeRef           |     SHORT |        2 |           | 0 200
           166 | 0x0016 GPSDestLongitude             |     SHORT |        4 |       794 | 0 0 6000 4000
           178 | 0x0017 GPSDestBearingRef            | UNDEFINED |        4 |           | ...
           190 | 0x0018 GPSDestBearing               | UNDEFINED |        4 |           | ...
           202 | 0x0019 GPSDestDistanceRef           | SRATIONAL |        1 |       802 | 0/6
           214 | 0x001b GPSProcessingMethod          |     SHORT |        7 |       810 | 0 6016 4016 6016 4016 ...
           226 | 0x001c GPSAreaInformation           | UNDEFINED |        3 |           | ...
           238 | 0x001d GPSDateStamp                 |     ASCII |        8 |       824 | 2567806
           250 | 0x001e GPSDifferential              |     SHORT |        1 |           | 1
           262 | 0x001f                              | UNDEFINED |        8 |       832 | 0100...
           274 | 0x0022                              |     SHORT |        1 |           | 65535
           286 | 0x0023                              | UNDEFINED |       58 |       840 | 0100STANDARD............STANDARD ...
           298 | 0x0024                              | UNDEFINED |        4 |           | ...
           310 | 0x0025                              | UNDEFINED |       14 |       898 | H.....H......
           322 | 0x002b                              | UNDEFINED |       16 |       912 | 0100...........
           334 | 0x002c                              | UNDEFINED |      574 |       928 | 0101#.......................... ...
           346 | 0x002d                              |     SHORT |        3 |      1502 | 512 0 0
           358 | 0x0032                              | UNDEFINED |        8 |      1508 | 0100...
           370 | 0x0035                              | UNDEFINED |        6 |      1516 | 0200.
           382 | 0x003b                              |  RATIONAL |        4 |      1522 | 256/256 256/256 256/256 256/256
           394 | 0x003c                              |     SHORT |        1 |           | 1
           406 | 0x0083                              |      BYTE |        1 |           | .
           418 | 0x0084                              |  RATIONAL |        4 |      1554 | 180/10 2500/10 35/10 63/10
           430 | 0x0087                              |      BYTE |        1 |           | 
           442 | 0x0089                              |     SHORT |        1 |           | 0
           454 | 0x008a                              |     SHORT |        1 |           | 1
           466 | 0x008b                              | UNDEFINED |        4 |           | 7..
           478 | 0x0095                              |     ASCII |        5 |      1586 | OFF 
           490 | 0x0097                              | UNDEFINED |     1188 |      1592 | 0219.dU....W..2......:.......F.# ...
           502 | 0x0098                              | UNDEFINED |       33 |      2780 | 0204.W....z.o..#[.....!o.x..E... ...
           514 | 0x009d                              |     SHORT |        1 |           | 0
           526 | 0x009e                              |     SHORT |       10 |      2814 | 0 0 0 0 0 ...
           538 | 0x00a2                              |      LONG |        1 |           | 6173648
           550 | 0x00a3                              |      BYTE |        1 |           | 
           562 | 0x00a7                              |      LONG |        1 |           | 9608
           574 | 0x00a8                              | UNDEFINED |       49 |      2834 | 0106........................... ...
           586 | 0x00ab                              |     ASCII |       16 |      2884 | AUTO(FLASH OFF)
           598 | 0x00b0                              | UNDEFINED |       16 |      2900 | 0100...........
           610 | 0x00b1                              |     SHORT |        1 |           | 4
           622 | 0x00b6                              | UNDEFINED |        8 |      2916 | .......
           634 | 0x00b7                              | UNDEFINED |       30 |      2924 | 0100....i....................
           646 | 0x00b8                              | UNDEFINED |      172 |      2954 | 0100..e........................ ...
           658 | 0x00bb                              | UNDEFINED |        8 |      3126 | 0200...
           670 | 0x00bf                              |     SHORT |        1 |           | 0
           682 | 0x00c0                              | UNDEFINED |        8 |      3134 | <......
      END MemIo
         428 | 0x9286 UserComment                  | UNDEFINED |       44 |      3972 | ASCII...                         ...
         440 | 0x9290 SubSecTime                   |     ASCII |        3 |           | 00
         452 | 0x9291 SubSecTimeOriginal           |     ASCII |        3 |           | 00
         464 | 0x9292 SubSecTimeDigitized          |     ASCII |        3 |           | 00
         476 | 0xa000 FlashpixVersion              | UNDEFINED |        4 |           | 0100
         488 | 0xa001 ColorSpace                   |     SHORT |        1 |           | 1
         500 | 0xa002 PixelXDimension              |     SHORT |        1 |           | 6000
         512 | 0xa003 PixelYDimension              |     SHORT |        1 |           | 4000
         524 | 0xa005 InteroperabilityTag          |      LONG |        1 |           | 4066
         536 | 0xa217 SensingMethod                |     SHORT |        1 |           | 2
         548 | 0xa300 FileSource                   | UNDEFINED |        1 |           | .
         560 | 0xa301 SceneType                    | UNDEFINED |        1 |           | .
         572 | 0xa302 CFAPattern                   | UNDEFINED |        8 |      4016 | ........
         584 | 0xa401 CustomRendered               |     SHORT |        1 |           | 0
         596 | 0xa402 ExposureMode                 |     SHORT |        1 |           | 0
         608 | 0xa403 WhiteBalance                 |     SHORT |        1 |           | 0
         620 | 0xa404 DigitalZoomRatio             |  RATIONAL |        1 |      4024 | 1/1
         632 | 0xa405 FocalLengthIn35mmFilm        |     SHORT |        1 |           | 66
         644 | 0xa406 SceneCaptureType             |     SHORT |        1 |           | 0
         656 | 0xa407 GainControl                  |     SHORT |        1 |           | 0
         668 | 0xa408 Contrast                     |     SHORT |        1 |           | 0
         680 | 0xa409 Saturation                   |     SHORT |        1 |           | 0
         692 | 0xa40a Sharpness                    |     SHORT |        1 |           | 0
         704 | 0xa40c SubjectDistanceRange         |     SHORT |        1 |           | 0
         716 | 0xa420 ImageUniqueID                |     ASCII |       33 |      4032 | 090caaf2c085f3e102513b24750041aa ...
    END MemIo
       130 | 0x8825 GPSTag                       |      LONG |        1 |           | 4096
      4346 | 0x0103 Compression                  |     SHORT |        1 |           | 6
      4358 | 0x011a XResolution                  |  RATIONAL |        1 |      4434 | 300/1
      4370 | 0x011b YResolution                  |  RATIONAL |        1 |      4442 | 300/1
      4382 | 0x0128 ResolutionUnit               |     SHORT |        1 |           | 2
      4394 | 0x0201 JPEGInterchangeFormat        |      LONG |        1 |           | 4450
      4406 | 0x0202 JPEGInterchangeFormatLength  |      LONG |        1 |           | 10837
      4418 | 0x0213 YCbCrPositioning             |     SHORT |        1 |           | 1
  END MemIo
   15305 | 0xffe1 APP1  |    2610 | http://ns.adobe.com/xap/1.0/.<?x
   17917 | 0xffed APP13 |      68 | Photoshop 3.0.8BIM.......'.....
  Record | DataSet | Name                     | Length | Data
       1 |       0 | ModelVersion             |      2 | ..
       1 |      90 | CharacterSet             |      3 | .%G
       2 |       0 | RecordVersion            |      2 | ..
       2 |     120 | Caption                  |     12 | Classic View
   17987 | 0xffd9 EOI  
543 rmills@rmillsmbp:~/temp $ 
```

clanmills commented:
@D4N I have no objections to this being merged.  Thank You for all your efforts on this issue - and everything else you deal with for Exiv2.  Thank You!

clanmills commented:
@frli8848 It's a good idea for the instructions to submit to GitHub to be easy to use.  However, working on that documentation mostly falls on the shoulders of @D4N - who is already over-loaded with many issues.  You'll appreciate that Dan has a full-time demanding job.

@Kicer86 is really good with git.  Perhaps you and Michal (his real name) could collaborate to polish CONTRIBUTING.md  Team Exiv2 use Slack to talk to each other informally.  If you send me an email (robin@clanmills.com), I'll send you an invitation to join Slack/Exiv2.

frli8848 commented:
I can work on a draft for the CONTRIBUTING.md file and an email is on the way :-)

D4N commented:
Thanks for that. You'll find my e-mail address in the commit log. Or you
can also submit a PR, too.

Fredrik Lingvall <notifications@github.com> writes:

> I can work on a draft for the CONTRIBUTING.md file and an email is on the way :-)
>
> -- 
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/227#issuecomment-367665816

clanmills commented:
@frli8848 I've sent you an invitation to join us on Slack.  Welcome on-board.  You are welcome to introduce yourself to everybody on the "# general" channel.

@a17r Andreas:  You are also welcome to join us on Slack.  If you wish to do so, please email me:  robin@clanmills.com

clanmills commented:
PR #259 deals with the latest data from Mr Rottmerhusen.

clanmills commented:
@DAN.  I believe the change is good.  If the merge/conflict isn't too painful, I'd appreciate this going into the product.

There are a copy of issues concerning Freddy's documentation which are open (or in vague state).  #301 and #234
